zcash_primitives/src/merkle_tree/incremental.rs

```diff
@@ -0,0 +1,341 @@
+//! Implementations of serialization and parsing for Orchard note commitment trees.
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::convert::TryFrom;
+use std::hash::Hash;
+use std::io::{self, Read, Write};
+
+use incrementalmerkletree::{
+    bridgetree::{
+        AuthFragment, BridgeTree, Checkpoint, Frontier, Leaf, MerkleBridge, NonEmptyFrontier,
+    },
+    Hashable, Position,
+};
+use orchard::tree::MerkleCrhOrchardOutput;
+
+use super::CommitmentTree;
+use crate::serialize::{Optional, Vector};
+
+pub const SER_V1: u8 = 1;
+
+/// A hashable node within a Merkle tree.
+pub trait HashSer {
+    /// Parses a node from the given byte source.
+    fn read<R: Read>(reader: R) -> io::Result<Self>
+    where
+        Self: Sized;
+
+    /// Serializes this node.
+    fn write<W: Write>(&self, writer: W) -> io::Result<()>;
+}
+
+pub fn read_frontier_v0<H: Hashable + super::Hashable, R: Read>(
```

Comment:
Is it worth combining some of the shared `*_v0`, `*_v1` methods into a trait?

Reply:
I don't think so. When would you want that constraint?